File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Tutorials/clr_fundamentals/samples/threading/exclusive/cs/Monitor.cs
using System;
using System.IO;
using System.Threading;


class App {
   static StreamWriter sw = new StreamWriter(new FileStream("Test.txt", FileMode.Create));
   const Int32 maxThreads = 10;
   static Int32 numAsyncOps = maxThreads;
   static AutoResetEvent asyncOpsAreDone = new AutoResetEvent(false);
   

   public static void Main() {
      for (Int32 threadNum = 0; threadNum < maxThreads; threadNum++) {
         ThreadPool.QueueUserWorkItem(new WaitCallback(AppendNumbers), threadNum * 100);
      }

      asyncOpsAreDone.WaitOne();
      sw.Close();

      Console.Write("Press Enter to close window...");
      Console.Read();
   }


   // The callback method's signature MUST match that of a System.Threading.TimerCallback 
   // delegate (it takes an Object parameter and returns void)
   static void AppendNumbers(Object state) {
      Console.WriteLine("Writing to file:  Test.txt");

      Int32 numStart = (Int32) state;
      lock (sw) {
         for (Int32 x = 0; x < 100; x++) {
            sw.WriteLine(x + numStart);
            Thread.Sleep(0);
         }
      }
      if (Interlocked.Decrement(ref numAsyncOps) == 0)
         asyncOpsAreDone.Set();
   }
}

File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Tutorials/clr_fundamentals/samples/threading/exclusive/vb/Monitor.vb
imports System
imports System.IO
imports System.Threading


class App
   Shared Dim sw As New StreamWriter(new FileStream("Test.txt", FileMode.Create))
   const maxThreads As Int32 = 10
   Shared Dim numAsyncOps As Int32 = maxThreads
   Shared Dim asyncOpsAreDone As New AutoResetEvent(false)
   

   Shared Sub Main()
      Dim threadNum As Int32
      for threadNum = 0 To (maxThreads-1)
         ThreadPool.QueueUserWorkItem(new WaitCallback(AddressOf AppendNumbers), threadNum * 100)
      next

      asyncOpsAreDone.WaitOne()
      sw.Close()

      Console.Write("Press Enter to close window...")
      Console.Read()
   End Sub


   ' The callback method's signature MUST match that of a System.Threading.TimerCallback 
   ' delegate (it takes an Object parameter and returns void)
   Shared Sub AppendNumbers(state As Object)
      Console.WriteLine("Writing to file: Test.txt")

      Dim numStart As Int32 = CType(state, Int32)
      SyncLock (sw) 
         Dim x As Int32
         for x = 0 To (100-1)
            sw.WriteLine(x + numStart)
            Thread.Sleep(0)
         next
      End SyncLock
      if (Interlocked.Decrement(numAsyncOps) = 0)
         asyncOpsAreDone.Set()
      End If
   End Sub
End Class